File: tp-5/set.hs
module Set 
    (Set,emptyS, addS, belongS, sizeS, removeS, unionS, setToList)
where
    data Set a = Set [a]

    emptyS :: Set a
    emptyS = (Set [])

    addS :: Eq a => a -> Set a -> Set a
    addS x (Set ys) = if(pertenece x ys)
                        then (Set ys)
                        else (Set (x:ys))

    belongS :: Eq a => a -> Set a -> Bool
    belongS x (Set ls)= pertenece x ls

    sizeS :: Eq a => Set a -> Int
    sizeS (Set ls) = longitud ls

    removeS :: Eq a => a -> Set a -> Set a
    removeS k (Set ls) = (Set (remove k ls))

    unionS :: Eq a => Set a -> Set a -> Set a
    unionS (Set (x:xs)) (Set ys) = addS x (unionS (Set xs) (Set ys))

    setToList :: Eq a => Set a -> [a]
    setToList (Set xs) = xs

    pertenece :: Eq a => a -> [a] -> Bool
    pertenece k [] = False
    pertenece k (n:ns) = (n==k) || (pertenece k ns)

    longitud :: [a] -> Int
    longitud [] = 0
    longitud (n : ns) = 1 + longitud ns

    remove :: Eq a => a -> [a] -> [a]
    remove k (x:xs) = if(k == x)
                        then xs
                        else x: (remove k xs)

File: tp-10/Set.cpp
#include <iostream>
#include "Set.h"
#include "LinkedList.cpp"

using namespace std;

struct NodoS {
int elem; // valor del nodo
NodoS* siguiente; // puntero al siguiente nodo
};
struct SetSt {
int cantidad; // cantidad de elementos diferentes
NodoS* primero; // puntero al primer nodo
};

Set emptyS(){
    SetSt* set = new SetSt;
    set->cantidad = 0;
    set->primero = NULL;
    return set;
}

bool isEmptyS(Set s){
    return s->cantidad == 0;
}

bool belongsS(int x, Set s){
    int counter = 1;
    bool belongs = false;
    NodoS* current = s->primero;
    while(counter<=s->cantidad && not(belongs)){
        belongs = current->elem == x;
        counter++;
        current = current->siguiente;
    }
    return belongs;
}

void AddS(int x, Set s){
    int counter = 1;
    NodoS* nuevo = new NodoS;
    nuevo->elem = x;
    nuevo->siguiente = NULL;
    NodoS* current = s->primero;
    while(counter<s->cantidad){
        counter++;
        current = current->siguiente;
    }
    current->siguiente = nuevo;
    s->cantidad++;
}

void RemoveS(int x, Set s){
    int counter = 1;
    NodoS* current = s->primero;
    while(counter<=s->cantidad){
        counter++;
        if(current->elem == x){
            NodoS* temp = current;
            current = current->siguiente;
            delete temp;
            s->cantidad--;
        }
        else{
            current = current->siguiente;
        }   
    }
}

int sizeS(Set s){
    return s->cantidad;
}

LinkedList setToList(Set s){
    int counter = 1;
    LinkedListSt* list = new LinkedListSt;
    NodoS* current = s->primero;
    while(counter<=s->cantidad){
        Snoc(list, current->elem);
        counter++;
        current = current->siguiente;
    }
    return list;
}

void DestroyS(Set s){
    delete s;
}



